lang/wasm_instance_cache.hh

```diff
@@ -0,0 +1,326 @@
+        : _timer([this] { return on_timer(); })
+        , _timer_period(timer_period)
+        , _max_size(size)
+    {
+        setup_metrics();
+        _timer.arm(_timer_period);
+    }
+
+private:
+    wasm_instance load(wasm::context& ctx) {
+        auto store = wasmtime::Store(ctx.engine_ptr->get());
+        auto linker = wasmtime::Linker(ctx.engine_ptr->get());
+        auto wasi_def = linker.define_wasi();
+        if (!wasi_def) {
+            throw wasm::exception(format("Setting up wasi failed: {}", wasi_def.err().message()));
+        }
+        auto cfg = wasmtime::WasiConfig();
+        auto set_cfg = store.context().set_wasi(std::move(cfg));
+        if (!set_cfg) {
+            throw wasm::exception(format("Setting up wasi failed: {}", set_cfg.err().message()));
+        }
+        auto instance_res = linker.instantiate(store, *ctx.module);
+        if (!instance_res) {
+            throw wasm::exception(format("Creating a wasm runtime instance failed: {}", instance_res.err().message()));
+        }
+        auto instance = instance_res.unwrap();
+        auto function_obj = instance.get(store, ctx.function_name);
+        if (!function_obj) {
+            throw wasm::exception(format("Function {} was not found in given wasm source code", ctx.function_name));
+        }
+        wasmtime::Func* func = std::get_if<wasmtime::Func>(&*function_obj);
+        if (!func) {
+            throw wasm::exception(format("Exported object {} is not a function", ctx.function_name));
+        }
+        auto memory_export = instance.get(store, "memory");
+        if (!memory_export) {
+            throw wasm::exception("memory export not found - please export `memory` in the wasm module");
+        }
+        auto memory = std::get<wasmtime::Memory>(*memory_export);
+
+        return wasm::wasm_instance{.store=std::move(store), .instance=std::move(instance), .func=std::move(*func), .memory=std::move(memory)};
+    }
+
+    // lru must not be empty, and its elements must refer to entries in _cache
+    void evict_lru() noexcept {
+        auto& entry = _lru.front();
+        _total_size -= entry.instance_size;
+        _lru.pop_front();
+        auto it = _cache.find(entry.key);
+        _cache.erase(it);
+    }
+
+    void on_timer() noexcept {
+        auto now = seastar::lowres_clock::now();
+        while (!_lru.empty() && _lru.front().timestamp + _timer_period < now) {
+            evict_lru();
+        }
+    }
+
+public:
+    seastar::future<value_type> get(const key_type& key, wasm::context& ctx) {
+        cache_key_type key_pair(seastar::current_scheduling_group(), key);
+        auto& i = _cache[key_pair];
+        if (!i.instance) {
+            ++shard_stats().cache_misses;
+            i.instance = seastar::make_lw_shared<wasm_instance>(load(ctx));
+            _lru.emplace_back(key_pair, seastar::lowres_clock::now(), i.instance->memory.size(i.instance->store));
```

Comment:
If this throws, we end up with a map entry but no lru entry.

You can use `auto undo = defer(...); ...; undo.cancel()` to undo changes to data structure so that exceptions don't cause us to become inconsistent.

Also, _cache[key_pair] is dangerous because it creates an empty instance. Better to emplace, it is more atomic.

Reply:
Fixed in the rebase, although I used `try_emplace` instead of `emplace` because `emplace` does not actually support constructors with no parameters